File: apps/ios/FlexTranslate/Sources/FlexTranslate/Audio/EnergyVad.swift
import Foundation

// VAD по умолчанию для A1 — без модели; настоящий Silero VAD (sherpa-onnx) подменяется
// за этим протоколом на этапе A2.
//
// Идея: считаем RMS-энергию кадра, нормируем в 0...1 относительно полной шкалы Int16
// и сравниваем с `energyThreshold`. Простой автомат на два состояния с дебаунсом-hangover
// не дёргается на паузах между словами и коротких всплесках:
//   - silence -> speech: нужно `minSpeechFrames` громких кадров подряд
//   - speech -> silence: нужно `minSilenceFrames` тихих кадров подряд
// VadEvent отдаём только на подтверждённом переходе.
final class EnergyVad: Vad {
    // Порог RMS (нормированный 0...1), выше которого кадр считается "громким".
    private let energyThreshold: Float
    // Сколько громких кадров подряд нужно, чтобы подтвердить начало речи.
    private let minSpeechFrames: Int
    // Сколько тихих кадров подряд нужно, чтобы подтвердить конец речи (hangover).
    private let minSilenceFrames: Int

    private var state: VadState = .silence
    // Счётчик кадров подряд, противоположных текущему состоянию — копим улики
    // к следующему переходу.
    private var transitionFrames = 0

    // Дефолты под 16 кГц / кадры ~1024 сэмпла (~64 мс). Порог консервативный,
    // дебаунс в несколько кадров перекрывает естественные паузы между словами.
    init(
        energyThreshold: Float = 0.012,
        minSpeechFrames: Int = 2,
        minSilenceFrames: Int = 8
    ) {
        self.energyThreshold = energyThreshold
        self.minSpeechFrames = max(1, minSpeechFrames)
        self.minSilenceFrames = max(1, minSilenceFrames)
    }

    var currentState: VadState { state }

    func accept(_ frame: AudioFrame) -> VadEvent? {
        let loud = Self.rms(of: frame.pcm16) >= energyThreshold

        switch state {
        case .silence:
            guard loud else {
                transitionFrames = 0
                return nil
            }
            transitionFrames += 1
            guard transitionFrames >= minSpeechFrames else { return nil }
            state = .speech
            transitionFrames = 0
            return .speechStart(frame.monotonicTsMs)

        case .speech:
            guard !loud else {
                transitionFrames = 0
                return nil
            }
            transitionFrames += 1
            guard transitionFrames >= minSilenceFrames else { return nil }
            state = .silence
            transitionFrames = 0
            return .speechEnd(frame.monotonicTsMs)
        }
    }

    func reset() {
        state = .silence
        transitionFrames = 0
    }

    // Полная шкала Int16. Берём 32768 (а не Int16.max=32767), чтобы максимально
    // отрицательный сэмпл (-32768) уложился в нормированный диапазон [-1, 1].
    private static let int16FullScale = 32768.0

    // RMS-амплитуда, нормированная по полной шкале Int16.
    // Для пустого кадра возвращает 0 (нет сэмплов -> считаем тишиной).
    private static func rms(of samples: [Int16]) -> Float {
        guard !samples.isEmpty else { return 0 }
        var sumSquares: Double = 0
        for sample in samples {
            let normalised = Double(sample) / int16FullScale
            sumSquares += normalised * normalised
        }
        let meanSquare = sumSquares / Double(samples.count)
        return Float(meanSquare.squareRoot())
    }
}

File: apps/android/app/src/main/java/dev/flextranslate/audio/EnergyVad.kt
// VAD по умолчанию для A1 — без модели. Настоящий Silero VAD (sherpa-onnx) — это гейченный свап G003/A2 за тем же интерфейсом Vad.
package dev.flextranslate.audio

import dev.flextranslate.foundation.AudioFrame
import kotlin.math.sqrt

/**
 * Детерминированный VAD по RMS-энергии.
 *
 * Кадр считаем «громким», когда его RMS-амплитуда выше [rmsThreshold]. Сырую громкость через
 * гистерезис превращаем в устойчивые переходы речь/тишина:
 *  - SILENCE → SPEECH требует громкого звука подряд минимум [minSpeechDurationMs] (отсекает
 *    короткие всплески вроде одного шумного кадра).
 *  - SPEECH → SILENCE требует тишины подряд минимум [minSilenceDurationMs] («hangover» —
 *    перекрывает естественные паузы между словами, чтобы речь не рубилась посреди фразы).
 *
 * Без модели, без аллокаций на кадр, без зависимости от устройства — полностью тестируется на JVM.
 */
class EnergyVad(
    private val rmsThreshold: Double = DEFAULT_RMS_THRESHOLD,
    private val minSpeechDurationMs: Long = DEFAULT_MIN_SPEECH_DURATION_MS,
    private val minSilenceDurationMs: Long = DEFAULT_MIN_SILENCE_DURATION_MS,
) : Vad {

    private var state: VadState = VadState.SILENCE

    /** Монотонная метка времени начала текущего «кандидатного» (противоположного состоянию) прогона, или null. */
    private var candidateRunStartMs: Long? = null

    val currentState: VadState get() = state

    override fun accept(frame: AudioFrame): VadEvent? {
        val loud = rmsOf(frame.pcm16) >= rmsThreshold
        val now = frame.monotonicTsMs
        return when (state) {
            VadState.SILENCE -> evaluateSilence(loud, now)
            VadState.SPEECH -> evaluateSpeech(loud, now)
        }
    }

    override fun reset() {
        state = VadState.SILENCE
        candidateRunStartMs = null
    }

    /** В тишине устойчивый громкий прогон переводит в SPEECH и шлёт [VadEvent.SpeechStart]. */
    private fun evaluateSilence(loud: Boolean, now: Long): VadEvent? {
        if (!loud) {
            candidateRunStartMs = null
            return null
        }
        val runStart = candidateRunStartMs ?: now.also { candidateRunStartMs = it }
        if (now - runStart < minSpeechDurationMs) return null
        state = VadState.SPEECH
        candidateRunStartMs = null
        return VadEvent.SpeechStart(monotonicTsMs = now)
    }

    /** Во время речи устойчивая тишина (hangover) переводит в SILENCE и шлёт [VadEvent.SpeechEnd]. */
    private fun evaluateSpeech(loud: Boolean, now: Long): VadEvent? {
        if (loud) {
            candidateRunStartMs = null
            return null
        }
        val runStart = candidateRunStartMs ?: now.also { candidateRunStartMs = it }
        if (now - runStart < minSilenceDurationMs) return null
        state = VadState.SILENCE
        candidateRunStartMs = null
        return VadEvent.SpeechEnd(monotonicTsMs = now)
    }

    private fun rmsOf(pcm16: ShortArray): Double {
        if (pcm16.isEmpty()) return 0.0
        var sumSquares = 0.0
        for (sample in pcm16) {
            val value = sample.toDouble()
            sumSquares += value * value
        }
        return sqrt(sumSquares / pcm16.size)
    }

    companion object {
        /** ~1.4% от полной шкалы 16 бит — выше шума тихой комнаты, ниже энергии речи. */
        const val DEFAULT_RMS_THRESHOLD: Double = 450.0

        /** Речь должна держаться столько, прежде чем объявим начало (отсев всплесков). */
        const val DEFAULT_MIN_SPEECH_DURATION_MS: Long = 120L

        /** Тишина должна держаться столько, прежде чем объявим конец (hangover между словами). */
        const val DEFAULT_MIN_SILENCE_DURATION_MS: Long = 400L
    }
}
